File: removeDuplicates.js

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

// k = 2 which means maxium 2 repetation is allowed for certain number
// Taking slow pointer which will maintain the expected output and fast pointer(i) which will iterate over an array of which will starts from index 1.
// Taking count variable which will count how many times certain number have appeared.
// Copy fast pointer value to slow pointer and move slow pointer only when count is less than or equal to k which is 2 in this example.
// return slow pointer at the end as it will hold the response.
/**
 * @param {number[]} nums
 * @return {number}
 */
 var removeDuplicates = function(nums) {
    const k = 2;
    let count = 1;
    let slow = 1;
    for(let i = 1; i < nums.length; i++) {
        if(nums[i] === nums[i-1]) count++;
        else count = 1;
        if(count <= k) {
            nums[slow] = nums[i];
            slow++;
        }
    }
    return slow;
};

File: removeDuplicates.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

/*

Approach:
We use the 2 pointer pattern for this problem. Since we're told that we have to
sort the array in place such that the first K elements are in sorted order with
each element occurring <= twice. For this, the first pointer will traverse
normally in the array, but the second pointer will keep track of the running
index of the last of the solution array, such that each element ultimately
doesn't have more than two occurrences.

*/

package main

func removeDuplicates(nums []int) int {
	if len(nums) <= 1 {
		return len(nums)
	}

	j := 1 // start at second position
	counter := 1
	for i := 1; i < len(nums); i++ {
		if nums[i] == nums[i-1] {
			counter++
		} else {
			counter = 1
		}
		if counter <= 2 {
			nums[j] = nums[i]
			j++
		}
	}

	return j
}
